{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import gzip\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "\n",
    "GTF_HEADER  = ['seqname', 'source', 'feature', 'start', 'end', 'score',\n",
    "               'strand', 'frame']\n",
    "R_SEMICOLON = re.compile(r'\\s*;\\s*')\n",
    "R_COMMA     = re.compile(r'\\s*,\\s*')\n",
    "R_KEYVALUE  = re.compile(r'(\\s+|\\s*=\\s*)')\n",
    "\n",
    "\n",
    "def dataframe(filename):\n",
    "    \"\"\"Open an optionally gzipped GTF file and return a pandas.DataFrame.\n",
    "    \"\"\"\n",
    "    # Each column is a list stored as a value in this dict.\n",
    "    result = defaultdict(list)\n",
    "\n",
    "    for i, line in enumerate(lines(filename)):\n",
    "        for key in line.keys():\n",
    "            # This key has not been seen yet, so set it to None for all\n",
    "            # previous lines.\n",
    "            if key not in result:\n",
    "                result[key] = [None] * i\n",
    "\n",
    "        # Ensure this row has some value for each column.\n",
    "        for key in result.keys():\n",
    "            result[key].append(line.get(key, None))\n",
    "\n",
    "    return pd.DataFrame(result)\n",
    "\n",
    "\n",
    "\n",
    "def lines(filename):\n",
    "    \"\"\"Open an optionally gzipped GTF file and generate a dict for each line.\n",
    "    \"\"\"\n",
    "    fn_open = gzip.open if filename.endswith('.gz') else open\n",
    "\n",
    "    with fn_open(filename) as fh:\n",
    "        for line in fh:\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            else:\n",
    "                yield parse(line)\n",
    "\n",
    "\n",
    "def parse(line):\n",
    "    \"\"\"Parse a single GTF line and return a dict.\n",
    "    \"\"\"\n",
    "    result = {}\n",
    "\n",
    "    fields = line.rstrip().split('\\t')\n",
    "\n",
    "    for i, col in enumerate(GTF_HEADER):\n",
    "        result[col] = _get_value(fields[i])\n",
    "\n",
    "    # INFO field consists of \"key1=value;key2=value;...\".\n",
    "    infos = [x for x in re.split(R_SEMICOLON, fields[8]) if x.strip()]\n",
    "\n",
    "    for i, info in enumerate(infos, 1):\n",
    "        # It should be key=\"value\".\n",
    "        try:\n",
    "            key, _, value = re.split(R_KEYVALUE, info, 1)\n",
    "        # But sometimes it is just \"value\".\n",
    "        except ValueError:\n",
    "            key = 'INFO{}'.format(i)\n",
    "            value = info\n",
    "        # Ignore the field if there is no value.\n",
    "        if value:\n",
    "            result[key] = _get_value(value)\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "def _get_value(value):\n",
    "    if not value:\n",
    "        return None\n",
    "\n",
    "    # Strip double and single quotes.\n",
    "    value = value.strip('\"\\'')\n",
    "\n",
    "    # Return a list if the value has a comma.\n",
    "    if ',' in value:\n",
    "        value = re.split(R_COMMA, value)\n",
    "    # These values are equivalent to None.\n",
    "    elif value in ['', '.', 'NA']:\n",
    "        return None\n",
    "\n",
    "    return value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gtf = dataframe(\"/projects/ps-yeolab/genomes/hg19/gencode_v19/gencode.v19.annotation.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>end</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>feature</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>score</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14412</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>11869</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>None</td>\n",
       "      <td>29806</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>14363</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>None</td>\n",
       "      <td>31109</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000243485.2</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>29554</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000243485.2</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>None</td>\n",
       "      <td>36081</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>34554</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>None</td>\n",
       "      <td>54936</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000268020.2</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>52473</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000268020.2</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ccdsid    end exon_id exon_number feature frame            gene_id  \\\n",
       "0     None  14412    None        None    gene  None  ENSG00000223972.4   \n",
       "21    None  29806    None        None    gene  None  ENSG00000227232.4   \n",
       "82    None  31109    None        None    gene  None  ENSG00000243485.2   \n",
       "92    None  36081    None        None    gene  None  ENSG00000237613.2   \n",
       "100   None  54936    None        None    gene  None  ENSG00000268020.2   \n",
       "\n",
       "      gene_name gene_status   gene_type       ...       score seqname  source  \\\n",
       "0       DDX11L1       KNOWN  pseudogene       ...        None    chr1  HAVANA   \n",
       "21       WASH7P       KNOWN  pseudogene       ...        None    chr1  HAVANA   \n",
       "82   MIR1302-11       NOVEL     lincRNA       ...        None    chr1  HAVANA   \n",
       "92      FAM138A       KNOWN     lincRNA       ...        None    chr1  HAVANA   \n",
       "100      OR4G4P       KNOWN  pseudogene       ...        None    chr1  HAVANA   \n",
       "\n",
       "     start strand   tag      transcript_id transcript_name transcript_status  \\\n",
       "0    11869      +  None  ENSG00000223972.4         DDX11L1             KNOWN   \n",
       "21   14363      -  None  ENSG00000227232.4          WASH7P             KNOWN   \n",
       "82   29554      +  None  ENSG00000243485.2      MIR1302-11             NOVEL   \n",
       "92   34554      -  None  ENSG00000237613.2         FAM138A             KNOWN   \n",
       "100  52473      +  None  ENSG00000268020.2          OR4G4P             KNOWN   \n",
       "\n",
       "    transcript_type  \n",
       "0        pseudogene  \n",
       "21       pseudogene  \n",
       "82          lincRNA  \n",
       "92          lincRNA  \n",
       "100      pseudogene  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.loc[gtf['feature'] == 'gene'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "gene_ids_to_name = pd.DataFrame(index = gtf['gene_id'].unique())\n",
    "gene_ids_to_name = gene_ids_to_name.join(gtf['gene_name'], on='gene_id', how='left')\n",
    "gene_ids_to_name.to_csv(\"/path/where/to/save/gene_ids_to_name_gencodev19.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [ENSG00000223972.4, ENSG00000227232.4, ENSG00000243485.2, ENSG00000237613.2, ENSG00000268020.2]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = pd.DataFrame(index = gtf.loc[gtf['feature'] == 'gene']['gene_id'])\n",
    "names.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>end</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>feature</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>score</th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>strand</th>\n",
       "      <th>tag</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>14412</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>11869</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000223972.4</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>None</td>\n",
       "      <td>29806</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>14363</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000227232.4</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>None</td>\n",
       "      <td>31109</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000243485.2</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>29554</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000243485.2</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>None</td>\n",
       "      <td>36081</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>34554</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>lincRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>None</td>\n",
       "      <td>54936</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000268020.2</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>52473</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000268020.2</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>pseudogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ccdsid    end exon_id exon_number feature frame            gene_id  \\\n",
       "0     None  14412    None        None    gene  None  ENSG00000223972.4   \n",
       "21    None  29806    None        None    gene  None  ENSG00000227232.4   \n",
       "82    None  31109    None        None    gene  None  ENSG00000243485.2   \n",
       "92    None  36081    None        None    gene  None  ENSG00000237613.2   \n",
       "100   None  54936    None        None    gene  None  ENSG00000268020.2   \n",
       "\n",
       "      gene_name gene_status   gene_type       ...       score seqname  source  \\\n",
       "0       DDX11L1       KNOWN  pseudogene       ...        None    chr1  HAVANA   \n",
       "21       WASH7P       KNOWN  pseudogene       ...        None    chr1  HAVANA   \n",
       "82   MIR1302-11       NOVEL     lincRNA       ...        None    chr1  HAVANA   \n",
       "92      FAM138A       KNOWN     lincRNA       ...        None    chr1  HAVANA   \n",
       "100      OR4G4P       KNOWN  pseudogene       ...        None    chr1  HAVANA   \n",
       "\n",
       "     start strand   tag      transcript_id transcript_name transcript_status  \\\n",
       "0    11869      +  None  ENSG00000223972.4         DDX11L1             KNOWN   \n",
       "21   14363      -  None  ENSG00000227232.4          WASH7P             KNOWN   \n",
       "82   29554      +  None  ENSG00000243485.2      MIR1302-11             NOVEL   \n",
       "92   34554      -  None  ENSG00000237613.2         FAM138A             KNOWN   \n",
       "100  52473      +  None  ENSG00000268020.2          OR4G4P             KNOWN   \n",
       "\n",
       "    transcript_type  \n",
       "0        pseudogene  \n",
       "21       pseudogene  \n",
       "82          lincRNA  \n",
       "92          lincRNA  \n",
       "100      pseudogene  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new = gtf.loc[gtf['feature'] == 'gene']\n",
    "new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "new.set_index(\"gene_id\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(57820, 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>MIR1302-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>OR4G4P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_name\n",
       "gene_id                      \n",
       "ENSG00000223972.4     DDX11L1\n",
       "ENSG00000227232.4      WASH7P\n",
       "ENSG00000243485.2  MIR1302-11\n",
       "ENSG00000237613.2     FAM138A\n",
       "ENSG00000268020.2      OR4G4P"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = names.join(new['gene_name'], how=\"left\")\n",
    "print names.shape\n",
    "names.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#save the result as a tab separated file\n",
    "names.to_csv(\"/directory/to/save/gene_ids_to_name.txt\", sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
